### PR TITLE
iotdb-client-go

### PR DESCRIPTION
**Pull request by manlge**

### CODE REVIEW

**Comment by manlge** [comment]
PTAL

**Comment by manlge** [comment]
PTAL @liutaohua

**Comment by liutaohua** [comment]
thank u for u contribution, I have noticed your pr.

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+all: generate
+
+generate:
+	@thrift -out . -gen go rpc.thrift
```
Comment (jixuan1989) [inline Makefile:21]: can we download the rpc.thrift file from remote address directly in the makefile? 
https://raw.githubusercontent.com/apache/iotdb/master/thrift/src/main/thrift/rpc.thrift

File: client/session.go [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+	"net"
+	"time"
+
+	"github.com/apache/iotdb-client-go/rpc"
+	"github.com/apache/thrift/lib/go/thrift"
+)
+
+const (
+	DEFAULT_TIME_ZONE  = "Asia/Shanghai"
+	DEFAULT_FETCH_SIZE = 1024
```
Comment (Alima777) [inline client/session.go:35]: Use camelcase instead of snake case.

File: Makefile (hunk 1 above)
Comment (manlge) [reply Makefile:21]: good idea

**Comment by jixuan1989** [comment]
dependency thrift should be claimed in LICENSE-binary file. 
see: https://github.com/apache/iotdb/blob/master/LICENSE-binary

File: client/rpcdataset.go [hunk 3]
```diff
@@ -0,0 +1,448 @@
...
+	columnCount                int
+	sessionId                  int64
+	queryId                    int64
+	lastReadWasNull            bool
+	rowsIndex                  int
+	queryDataSet               *rpc.TSQueryDataSet
+	sql                        string
+	fetchSize                  int32
+	columnNameList             []string
+	columnTypeList             []int32
```
Comment (JackieTien97) [inline client/rpcdataset.go:49]: []int16

File: client/rpcdataset.go [hunk 4]
```diff
@@ -0,0 +1,448 @@
...
+	columnOrdinalMap           map[string]int32
+	columnTypeDeduplicatedList []int32
+	columnNameIndexMap         map[string]int32
+	columnTypeMap              map[string]int32
```
Comment (JackieTien97) [inline client/rpcdataset.go:53]: These two are useless, we can use columnOrdinalMap and columnTypeDeduplicatedList to infer the data type

File: client/rpcdataset.go [hunk 5]
```diff
@@ -0,0 +1,448 @@
...
+	queryId int64, client *rpc.TSIServiceClient, sessionId int64, queryDataSet *rpc.TSQueryDataSet,
+	ignoreTimeStamp bool) *IoTDBRpcDataSet {
+	typeMap := map[string]int32{
+		"BOOLEAN": BOOLEAN,
+		"INT32":   INT32,
+		"INT64":   INT64,
+		"FLOAT":   FLOAT,
+		"DOUBLE":  DOUBLE,
+		"TEXT":    TEXT,
+	}
```
Comment (JackieTien97) [inline client/rpcdataset.go:393]: shouldn't be a local var, may be a const value

File: client/sessiondataset.go [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+ * under the License.
+ */
+
+package client
+
+import "github.com/apache/iotdb-client-go/rpc"
+
+type SessionDataSet struct {
+	ioTDBRpcDataSet *IoTDBRpcDataSet
+}
```
Comment (JackieTien97) [inline client/sessiondataset.go:26]: Actually, we may only need one DataSet in go-client. For java, we need to support two kinds of dataset which are JDBCDataSet and SessionDataset, so we put the common part into IoTDBRpcDataSet. However, in go-client, we only need to support one.

File: client/rpcdataset.go [hunk 7]
```diff
@@ -0,0 +1,448 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package client
```
Comment (JackieTien97) [inline client/rpcdataset.go:20]: we should support close function in the dataset, or will cause resource leak.

File: client/session.go (hunk 2 above)
Comment (JackieTien97) [reply client/session.go:37]: +1

File: client/session.go [hunk 8]
```diff
@@ -0,0 +1,365 @@
...
+ *params
+ *path: string, complete time series path (starts from root)
+ *dataType: int32, data type for this time series
+ *encoding: int32, data type for this time series
+ *compressor: int32, compressing type for this time series
+ *
+ *return
+ *error: correctness of operation
+ */
+func (s *Session) CreateTimeseries(path string, dataType int32, encoding int32, compressor int32, attributes map[string]string, tags map[string]string) (r *rpc.TSStatus, err error) {
```
Comment (JackieTien97) [inline client/session.go:168]: also the type problem, change them to int16

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: client/sessiondataset.go (hunk 6 above)
Comment (manlge) [reply client/sessiondataset.go:28]: Thank you for your suggestion, IoTDBRpcDataSet is only used in the package, I think we can keep the current structure, maybe in the future we can also adapt to the golang's sql driver.

File: client/rpcdataset.go (hunk 4 above)
Comment (manlge) [reply client/rpcdataset.go]: PTAL https://github.com/apache/iotdb-client-go/pull/3/commits/1c2a0a3d62010d8b7938560ea4fc8c83ad1ce87d

File: client/field.go [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package client
+
+type Field struct {
+	dataType int32
```
[the remark this answers is not in the document]
Comment (manlge) [reply client/field.go]: PTAL https://github.com/apache/iotdb-client-go/pull/3/commits/f2c24f278629dc488c3372a6b2921d82f3b3e74a

File: client/rpcdataset.go (hunk 7 above)
Comment (manlge) [reply client/rpcdataset.go:20]: PTAL https://github.com/apache/iotdb-client-go/pull/3/commits/243db494cf3406a6ca7c1cda8b53440d6352c1c8

File: client/rpcdataset.go [hunk 10]
```diff
@@ -0,0 +1,545 @@
...
+		s.queryDataSet = resp.GetQueryDataSet()
+	}
+	return resp.HasResultSet, nil
+}
+
+func (s *IoTDBRpcDataSet) IsClosed() bool {
+	return s.closed
+}
+
+func (s *IoTDBRpcDataSet) Close() error {
```
Comment (JackieTien97) [inline client/rpcdataset.go:465]: The close function should send a close request to server to release the resource in server. You can refer to the close function in `IoTDBRpcDataSet` class in java

**Review by JackieTien97 — CHANGES_REQUESTED** [review]

File: client/rpcdataset.go (hunk 10 above)
Comment (manlge) [reply client/rpcdataset.go]: PTAL

**Review by JackieTien97 — APPROVED** [review]

**Pull request merged**